COSE envelope signature spec updates (#148)
notaryproject/notaryproject: signature-envelope-cose.md

# COSE Sign1 Envelope Serialization

In COSE ([rfc8152](https://datatracker.ietf.org/doc/html/rfc8152)), data is stored as either payload or headers (protected and unprotected).
Notary v2 supports [COSE_Sign1_Tagged](https://datatracker.ietf.org/doc/html/rfc8152#section-4.2) as a signature envelope with some additional constraints on the header fields.

> A JWS based signature will be persisted with a blob `mediaType` of `"application/cose"`

Reviewer: Typo: I believe this should be COSE instead of JWS on this line

    }
}
```

Unless explicitly specified as OPTIONAL, all fields are required.

**Payload**: COSE signs the payload as defined in the [Payload](#payload) section. Detached payloads are not supported.

**ProtectedHeaders**: Notary v2 supports the following protected headers. Other header fields can be included but will be ignored.

```
{
  / crit / 2: [
    / cty / 3,
    'signingtime',
    'expiry'
  ],
  / cty / 3: 'application/vnd.cncf.oras.artifact.descriptor.v1+json',
  'signingtime': 1234567891000,
  'expiry': 1234567891011
}
```

Note: The above example is represented using the [extended CBOR diagnostic notation](https://datatracker.ietf.org/doc/html/rfc8152#appendix-C).

- **`crit`** (*array of integers or strings*): This REQUIRED property (label: `2`) lists the headers that implementation MUST understand and process.
  The array MUST contain `3` (`cty`), and `signingtime`. If `expiry` is presented, the array MUST also contain `expiry`.
- **`cty`** (*string*): The REQUIRED property content-type (label: `3`) is used to declare the media type of the secured content (the payload).

> - **`signingtime`** (*datetime*): The REQUIRED property identifies the time at which the signature was generated.

Reviewer 2: Should be more specific on the format. `*datetime*` isn't defined anywhere.

Author: @letmaik, which RFC should we reference here?

Reviewer 3: RFC 7519 specifies this field as number of _seconds_ since epoch, if it's to be aligned with formats used in JWT.

Reviewer 4: Its value is the number of seconds from `1970-01-01T00:00Z` in UTC time, commonly known as UNIX timestamp

> - **`signingtime`** (*datetime*): The REQUIRED property identifies the time at which the signature was generated.

Reviewer 2: Any reason this can't use the JWT name `"iat"`? Same for expiry `"exp"`.

Author: Happy to change to whatever folks prefer

Reviewer 3: Is this the same as the difference between iat and nbf?  The signature could be generated at one time but not be considered valid until its issued?

Reviewer 4: We're consistently using io.cncf.notary.expiry on both and io.cncf.notary.signingTime

- **`expiry`** (*datetime*): This OPTIONAL property contains the expiration time on or after which the signature must not be considered valid.

**UnprotectedHeaders**: Notary v2 supports two unprotected headers: `timestamp` and `x5chain`.

```
{
  'timestamp': << TimeStampToken >>,
  / x5chain / 33: [
    << DER(leafCert) >>,
    << DER(intermediate CACert) >>,
    << DER(rootCert) >>
  ]
}
```

Note: `<<` and `>>` are used to notate the byte string resulting from encoding the data item.

> - **`timestamp`** (*byte string*): This OPTIONAL property is used to store time stamp token.

Reviewer 2: Should be more specific on format, i.e. DER encoding.

Author: Can you please propose what you'd like?

Reviewer 4: Its value is the number of seconds from `1970-01-01T00:00Z` in UTC time, commonly known as UNIX timestamp

- **`x5chain`** (*array of byte strings*): This REQUIRED property (label: `33` by [IANA](https://www.iana.org/assignments/cose/cose.xhtml#header-parameters)) contains the list of X.509 certificate or certificate chain ([RFC5280](https://datatracker.ietf.org/doc/html/rfc5280)) corresponding to the key used to digitally sign the COSE.
  The certificate containing the public key corresponding to the key used to digitally sign the COSE MUST be the first certificate.
  Optionally, this header can be presented in the protected header.

**Signature**: In COSE, signature is calculated by constructing the `Sig_structure` for `COSE_Sign1`.
The process is described below:

1. Encode the protected header into a CBOR object as a byte string named `body_protected`.
2. Construct the `Sig_structure` for `COSE_Sign1`.
    ```
    Sig_structure = [
        / context / 'Signature1',
        / body_protected / << ProtectedHeaders >>,
        / external_aad / h'',
        / payload / << Payload >>,
    ]
    ```
3. Encode `Sig_structure` into a CBOR object as a byte string named `ToBeSigned`.

> 4. Compute the signature on the `ToBeSigned` constructed in the previous step by using the signature algorithm defined in the corresponding header element: `alg`.

Reviewer 2: `alg` is missing in the protected headers above.

Reviewer 4: This is present in the spec

   This is the value of the signature property used in the signature envelope.

**Signature Envelope**: The final signature envelope is a `COSE_Sign1_Tagged` object, consisting of Payload, ProtectedHeaders, UnprotectedHeaders, and Signature.

```
18(
  [
    / protected / << {
      / crit / 2: [

>         / cty / 3,

Reviewer 2: COSE spec says "Integer labels in the range of 0 to 7 SHOULD be omitted.". This is less strict than JWS which forbids it, but would probably do the same for consistency.

# COSE Sign1 Signature Envelope

This specification implements the [Notary v2 Signature specification](signature-specification.md) using 
CBOR Object Signing and Encryption (COSE). COSE ([RFC8152](https://datatracker.ietf.org/doc/html/rfc8152)) is a CBOR based envelope format for digital signatures over any type of payload (e.g. CBOR, JSON, binary).
Notary v2 specifically supports [COSE_Sign1_Tagged](https://datatracker.ietf.org/doc/html/rfc8152#section-4.2) as a signature envelope.

## Storage

> A COSE signature envelope will be stored in an OCI registry as a blob, and referenced in the signature manifest as a blob with `mediaType` of `"application/cose"`.

Reviewer 5: a more precise media type to use in this context would be `application/cose; cose-type="cose-sign1"` (see [Section 2 of STD 96](https://www.rfc-editor.org/rfc/rfc9052#section-2)).

Reviewer 2: I think this is redundant since the COSE envelopes are tagged anyway. In my mind this would make client code harder to write because now you need to define how media type parameters are encoded in this context. This is not HTTP (which has its own definition for encoding those).

Reviewer 5: > I think this is redundant since the COSE envelopes are tagged anyway.

Yeah it's redundant.  Whether this hint is worthwhile depends on how processing is implemented.  With the extra information, the COSE payload can be dispatched directly to the right API, whilst without, the implementation needs to peek into the COSE tag to dispatch.

> In my mind this would make client code harder to write because now you need to define how media type parameters are encoded in this context. This is not HTTP (which has its own definition for encoding those).

Not sure I understand what the encoding problem is?  The media parameters that'd be used here are all UTF-8 strings (which both JSON and I-JSON support off-the-shelf) with a-priori known values.

( I have the feeling I am missing something though :-) )

Reviewer 2: In practice, even outside the context of Email or HTTP headers, people use the same syntax as first defined in https://datatracker.ietf.org/doc/html/rfc2045#section-5.1 to encode parameters, especially for protocols/formats where you want to use a single string to encode the media type. 
But it has to be made explicit with a reference that this is what's happening here.
From an implementation point of view, for example in Python, clients would have to do the following to match the media type (note the `parse_header` function name):
```py
import cgi
mediatype, params = cgi.parse_header('application/cose; cose-type="cose-sign1"')
if mediatype == "application/cose" and params["cose-type"] == "cose-sign1":
  # process
```
This is necessary since whitespace becomes a factor now. Previously, a simple string comparison would have been enough: `mediatype == "application/cose"`.
Of course you could prescribe that the media type string must always use a specific fixed string that can be matched by string comparison. There is some chance that people will mess it up though.

I think, given that the ORAS artifact type already says that it is a (single) signature (`"artifactType": "application/vnd.cncf.notary.signature"`), I don't see why it has to be repeated here using the parameter.

Reviewer 4: Agree with letmaik - going to close this comment out.

Signature Manifest Example

```jsonc
{
    "mediaType": "application/vnd.cncf.oras.artifact.manifest.v1+json",
    "artifactType": "application/vnd.cncf.notary.signature",
    "blobs": [
        {
            "mediaType": "application/cose",
            "digest": "sha256:9834876dcfb05cb167a5c24953eba58c4ac89b1adf57f28f2f9d09af107ee8f0",
            "size": 32654
        }
    ],
    "subject": {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "digest": "sha256:73c803930ea3ba1e54bc25c2bdc53edd0284c62ed651fe7b00369da519a3c333",
        "size": 16724
    },
    "annotations": {
        "io.cncf.notary.x509chain.thumbprint#S256": 
        "[\"B7A69A70992AE4F9FF103EBE04A2C3BA6C777E439253CE36562E6E98375068C3\",\"932EB6F5598435D4EF23F97B0B5ACB515FAE2B8D8FAC046AB813DDC419DD5E89\"]"
    }
}
```

## COSE Payload

> The COSE envelope contains a [Notary v2 Payload](./signature-specification.md#payload).

Reviewer 6: Would be good to add an example of Notary V2 payload so reader does not have to jump to generic signature spec to see one.

```suggestion
The COSE envelope contains a [Notary v2 Payload](./signature-specification.md#payload).

Example of Notary v2 payload

```jsonc
{
  "targetArtifact": {
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "digest": "sha256:73c803930ea3ba1e54bc25c2bdc53edd0284c62ed651fe7b00369da519a3c333",
    "size": 16724,
    "annotations": {
        "io.wabbit-networks.buildId": "123"  // user defined metadata
    }
  }
}
```

```

## Protected Header

The COSE envelope for Notary v2 uses the following header parameters:

- [Common parameters](https://www.iana.org/assignments/cose/cose.xhtml#header-parameters)

>   - Label `1`: `alg`

Reviewer 6: @shizhMSFT , is `alg` a required header for COSE? We included it in JWS envelope as the JWS spec requires it. If not, can it be ommited and use signature algorithm from signing key/cert as source of truth?

Reviewer 7: According to the COSE spec, `alg` header MUST be present if there is no externally supplied data.

Reviewer 7: In the notary scenario, there is no externally supplied data. Therefore, `alg` is a required header.

Reviewer 6: Could you clarify that a bit more, what does externally supplied data mean in this context? Isn't signing cert's signature alg from certificate chain enough in this case?

Reviewer 8: The externally supplied data is explained in COSE spec, see https://www.rfc-editor.org/rfc/rfc9052.html#Extern_AAD, and Notary v2 doesn't use it. The value of `alg` is explained here https://www.rfc-editor.org/rfc/rfc9052.html#name-common-cose-header-paramete, which was copied below for your convenience:
"alg: This header parameter is used to indicate the algorithm used for the security processing. This header parameter MUST be authenticated where the ability to do so exists. This support is provided by AEAD algorithms or construction (e.g., COSE_Sign and COSE_Mac0). This authentication can be done either by placing the header parameter in the protected-header-parameters bucket or as part of the externally supplied data ([Section 4.3](https://www.rfc-editor.org/rfc/rfc9052.html#Extern_AAD)). The value is taken from the "COSE Algorithms" registry (see [[COSE.Algorithms](https://www.rfc-editor.org/rfc/rfc9052.html#COSE.Algorithms)])."

  - Label `2`: `crit`
  - Label `3`: `content type`
- Additional parameters with collision resistant names.
  - `io.cncf.notary.signingScheme`
  - `io.cncf.notary.signingTime`
  - `io.cncf.notary.authenticSigningTime`
  - `io.cncf.notary.expiry`

> Example with Signing Scheme `notary.x509`

Reviewer 6: ```suggestion
COSE examples in this document use [CBOR diagnostic notation](https://www.rfc-editor.org/rfc/rfc7049#section-6)([examples](https://datatracker.ietf.org/doc/html/rfc8152#appendix-C)).

Example with Signing Scheme `notary.x509`
```

Reviewer 8: @gokarnm  There is already a sentence after the example, see below, so I don't think we need another note before the example.

" Note: The above examples are represented using the [extended CBOR diagnostic notation](https://datatracker.ietf.org/doc/html/rfc8152#appendix-C)."

Reviewer 6: Yes that helps, maybe I missed it while reviewing last version. Resolving.

```yaml
{
  / alg / 1: / PS384 / -38,
  / crit / 2: [
    'io.cncf.notary.signingScheme',
    'io.cncf.notary.expiry'
  ],

>   / cty / 3: 'application/vnd.cncf.notary.payload.v1+json',

Reviewer 9: this header struct looks good to me.

- **[`crit`](https://datatracker.ietf.org/doc/html/rfc8152#section-3.1)** (*array of int/tstr*): This REQUIRED parameter (label `2`) lists the header parameters that implementations MUST understand and process. It MUST only contain parameters apart from integer labels in the range of 0 to 8. This header MUST contain `io.cncf.notary.signingScheme` which is a required critical header, and optionally contain `io.cncf.notary.authenticSigningTime` and `io.cncf.notary.expiry` if these critical headers are present in the signature.
- **[`content type`](https://datatracker.ietf.org/doc/html/rfc8152#section-3.1)** (*tstr*): The REQUIRED parameter content type (label `3`) is used to declare the media type of the secured content (the payload). The supported value is `application/vnd.cncf.notary.payload.v1+json`.
- **`io.cncf.notary.signingScheme`** (*tstr*, critical): This REQUIRED header specifies the [Notary v2 Signing Scheme](./signing-scheme.md) used by the signature. Supported values are `notary.x509` and `notary.x509.signingAuthority`.

> - **`io.cncf.notary.signingTime`** (*date/time*): This header specifies the time at which the signature was generated. This is an untrusted date/time, and therefore not used in trust decisions. Its value is an Epoch-Based Date/Time defined in [RFC 8949](https://datatracker.ietf.org/doc/html/rfc8949#section-3.4.2). This claim is REQUIRED and only valid when signing scheme is `notary.x509`.

Reviewer 6: @shizhMSFT  Is there a reason the RFC 3339 text based date time representation supported by COSE (tag 0) is not used in favor of epoch based representation (tag 1)? More concise representation may be one reason (~ 8 vs 20 bytes), are there other stronger reasons?

Reviewer 7: COSE is based on CBOR where CBOR standards for **Concise Binary** Object Representation. Concise by default is the principle.

Oppositely, JWS is based on JSON, which is text-based not binary-based. That's why RFC3339 is preferable for JWS.

- **`io.cncf.notary.authenticSigningTime`** (*date/time*, critical): This header specifies the authenticated time at which the signature was generated. Its value is an Epoch-Based Date/Time defined in [RFC 8949](https://datatracker.ietf.org/doc/html/rfc8949#section-3.4.2). This claim is REQUIRED and only valid when signing scheme is `notary.x509.signingAuthority` .
- **`io.cncf.notary.expiry`** (*date/time*, critical): This OPTIONAL header provides a "best by use" time for the artifact, as defined by the signer. Its value is an Epoch-Based Date/Time defined in [RFC 8949](https://datatracker.ietf.org/doc/html/rfc8949#section-3.4.2).

## Unprotected Headers

Notary v2 supports the following unprotected header parameters:

- `io.cncf.notary.timestampSignature`
- Label `33`: `x5chain`
- `io.cncf.notary.signingAgent`

```yaml
{
  / x5chain / 33: [
    << DER(leafCert) >>,

>     << DER(intermediate CACert) >>,

Reviewer 9: I am still a bit nervous about exactly what this means, but the approach seems reasonable... I would just want to see a complete example.

    << DER(rootCert) >>
  ],
  'io.cncf.notary.timestampSignature': << TimeStampToken >>,
  'io.cncf.notary.signingAgent': 'notation/1.0.0'
}
```

> Note: `<<` and `>>` are used to notate the byte string resulting from encoding the data item.

Reviewer 5: ```suggestion
Note: `<<` and `>>` are used to notate the CBOR byte string resulting from encoding the data item.
```

> - **[`x5chain`](https://datatracker.ietf.org/doc/html/draft-ietf-cose-x509-08#section-2)** (*array of bstr*): This REQUIRED parameter (label `33` by [IANA](https://www.iana.org/assignments/cose/cose.xhtml#header-parameters)) contains the ordered list of X.509 certificate or certificate chain ([RFC5280](https://datatracker.ietf.org/doc/html/rfc5280)) corresponding to the key used to digitally sign the COSE. The certificate chain is represented as an array of certificate value byte string, each certificate in the array is DER encoded. The certificate containing the public key corresponding to the key used to digitally sign the COSE MUST be the first certificate, followed by the intermediate and root certificates in the correct order. Refer [*Certificate Chain* unsigned attribute](signature-specification.md#unsigned-attributes) for more details. Optionally, this header can be presented in the protected header.

Reviewer 5: ```suggestion
- **[`x5chain`](https://datatracker.ietf.org/doc/html/draft-ietf-cose-x509-08#section-2)** (*array of bstr*): This REQUIRED parameter (label `33` by [IANA](https://www.iana.org/assignments/cose/cose.xhtml#header-parameters)) contains the ordered list of X.509 certificate or certificate chain ([RFC5280](https://datatracker.ietf.org/doc/html/rfc5280)) corresponding to the key used to digitally sign the COSE. The certificate chain is represented as an array of certificate, each certificate in the array is DER encoded and then wrapped in a CBOR byte string. The certificate containing the public key corresponding to the key used to digitally sign the COSE MUST be the first certificate, followed by the intermediate and root certificates in the correct order. Refer [*Certificate Chain* unsigned attribute](signature-specification.md#unsigned-attributes) for more details. Optionally, this header can be presented in the protected header.
```

>   - **TODO** update signature specification to allow chains in protected header

Reviewer 6: What does the implementation use now?

Reviewer 7: Currently, it is the same as JWS that chains are from unprotected header.

See https://github.com/notaryproject/notation-core-go/blob/v0.2.0-beta.1/signature/cose/envelope.go#L218

- **`io.cncf.notary.timestampSignature`** (*bstr*): This OPTIONAL header is used to store countersignature that provides authentic signing time. Only [RFC3161]([rfc3161](https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.2)) compliant `TimeStampToken` are supported.
  - **TODO** Define the opaque datum (hash of envelope) that is sent to TSA, and how TSA response (time stamp token) is represented in this header.
- **`io.cncf.notary.signingAgent`** (*tstr*): This OPTIONAL header provides the identifier of a client (e.g. Notation) that produced the signature. E.g. `notation/1.0.0`. Refer [*Signing Agent* unsigned attribute](signature-specification.md#unsigned-attributes) for more details.

## Signature

In COSE, signature is calculated by constructing the `Sig_structure` for `COSE_Sign1`.
The process is described below:

1. Encode the protected header into a CBOR object as a byte string named `body_protected`.
2. Construct the `Sig_structure` for `COSE_Sign1`.
    ```yaml
    Sig_structure = [

>         / context / 'Signature1',

Reviewer 6: It'd be good to add detail about `context Signature1`, that the identifying context for COSE_Sign1 is `"Signature1"`.

Reviewer 7: The context `Signature1` is fixed for `COSE_Sign1` and is defined by the COSE spec. No one can change it.

Reviewer 6: Ya, I meant just documenting what `context Signature1` is in this section.

Reviewer 8: @gokarnm See below my proposal for the 2nd bullet, what do you think?
from 
"Construct the `Sig_structure` for `COSE_Sign1`." 
to 
"Construct the `Sig_structure` for `COSE_Sign1`. The context string `Signature1` is used to identify the context of signatures using the `COSE_Sign1` structure."

Reviewer 6: @yizha1 how about?

"Construct the Sig_structure for `COSE_Sign1`.  The `context` field is used to identify the context of signature,   string `Signature1` is used to identity a `COSE_Sign1` structure."

        / body_protected / << ProtectedHeaders >>,

>         / external_aad / h'',

Reviewer 6: What is `external_aad` , it's not defined here.

Reviewer 7: The `external_aad` is the COSE external data.

Reviewer 8: @gokarnm  Did shiwei answer your questions?

Reviewer 6: Some content describing this field and if/how Notary v2 uses it, and any example will help.

Reviewer 8: This field is explained here https://www.rfc-editor.org/rfc/rfc9052.html#Extern_AAD, since Notary v2 doesn't use it, so it's empty.

        / payload / << Payload >>,
    ]
    ```
3. Encode `Sig_structure` into a CBOR object as a byte string named `ToBeSigned`.

> 4. Compute the signature on the `ToBeSigned` constructed in the previous step by using the signature algorithm defined in the corresponding header element: `alg`.

Reviewer 6: ```suggestion
4. Compute the signature on the `ToBeSigned` constructed in the previous step by using the signature algorithm of the signing key, which MUST match the corresponding protected header element `alg`.
```

        << DER(rootCert) >>
      ],
      'io.cncf.notary.timestampSignature': << TimeStampToken >>,
      'io.cncf.notary.signingAgent': 'notation/1.0.0'
    },
    / payload / << descriptor >>,
    / signature / << sign( << Sig_structure >> ) >>
  ]
)
```

## Implementation Constraints

### Supported `alg` header values

Notary v2 implementation MUST enforce the following constraints on signature generation and verification:

1. `alg` parameter value MUST NOT be a symmetric-key algorithm such as `HMAC`.
1. `alg` parameter value MUST be same as that of signature algorithm identified using signing certificate's public key algorithm and size.
1. `alg` parameter values for various signature algorithms is a subset of values supported by [COSE](https://www.iana.org/assignments/cose/cose.xhtml#algorithms).

**Mapping of Notary v2 approved algorithms to COSE `alg` header parameter values**

  | Signature Algorithm             | `alg` Label       |
  | ------------------------------- | ----------------- |
  | RSASSA-PSS with SHA-256         | -37               |
  | RSASSA-PSS with SHA-384         | -38               |
  | RSASSA-PSS with SHA-512         | -39               |

>   | ECDSA on secp256r1 with SHA-256 | -7                |

Reviewer 9: I assume lack of EdDSA is intentional here.
